### ISSUE TITLE
geo_em variables depend on LU_INDEX, problem?

### ISSUE THREAD

**Issue by matthiasdemuzere** [matthiasdemuzere/wrf-lcz-KL#4]
Hi all, 

I was chatting with Oscar about his WUDAPT-to-WRF developments, to make sure we are not doing the same thing at the same time. Seems not, as he does not intend to really provide a new tool / skip the intermediate txt/binary steps.

One thing he did mention is the following:

> One thing to keep in mind – if I remember well – when directly implementing the required info into whether geo_em or wrfinput files is that some geo_em variables (e.g. LAI12M, ALBEDO12M or GREENFRAC) are calculated based on the LU_INDEX and the different binary files that are given. I may be wrong, but I read that it’s always best to implement the variables via the geogrid binaries instead of changing the NetCDF input files directly. The model will run anyhow, but some surface parameters may have some small discrepancies. Most likely in places where urban data are poor – I expect less effect on a city like London or Berlin for example.

Is this posing a problem for what we have in mind? 
Or is this what you previously referred to @andreazonato / @dargueso?

If a problem, any suggestion on how to cope with this? Or perhaps address only when we have most of the code in place?

**Comment by andreazonato** [comment]
@matthiasdemuzere 
I think the discrepancies whose we can get in terms of GREENFRAC, for example, are not really an issue in the urban areas. Here They are not read into the urban scheme routine, so they are not important.

They can become important for the experiment A), since we change the urban landuse with croplands, but anyhow I think there is no difference for this variables between urban and rural cells, wee just simply have to check.

**Comment by matthiasdemuzere** [comment]
Ok, I had a quick look at some of these fields. I am displaying there temporal mean (as they are monthly) for the urban areas (LU_FRACTION == 13) only (left) and the whole domain (right).

GREENFRAC:
![LU_INDEX_GREENFRAC](https://user-images.githubusercontent.com/6926916/114894193-b6e31c00-9e0e-11eb-866b-acb41e4f8e44.jpg)

ALBEDO12M:
![LU_INDEX_ALBEDO12M](https://user-images.githubusercontent.com/6926916/114894349-da0dcb80-9e0e-11eb-8d7c-d416e1b8408e.jpg)

LAI12M:
![LU_INDEX_LAI12M](https://user-images.githubusercontent.com/6926916/114894465-f0b42280-9e0e-11eb-8d62-9679ed2337d0.jpg)

And then there is LANDUSEF, which is Noah-modified 21-category IGBP-MODIS landuse fractions.
Position 13 reflects the urban area, this position has all ones:
![LU_INDEX_LANDUSEF](https://user-images.githubusercontent.com/6926916/114896637-ccf1dc00-9e10-11eb-819e-9fead88d8ae2.jpg)

So... what I learn from the above:
* GREENFRAC seems loosely connected to the urban area. Yet low GREENFRAC values also occur outside of this urban area?
* ALBEDO and LAI in my opinion seem not directly linked to the urban area?
* I assume that LANDUSEF should be changed simultaneously with LU_INDEX when we replace the urban extent?

Yet this is all derived visually from the geom_em*.nc file.

Someone should now how this information is compiled, and what parameters are relevant no?
We should try to figure this out: if we want to make a package that is widely used, it should function correctly ...

That's my two cents. Happy to hear your opinion @dargueso @andreazonato. And ideally also suggestions how to fix this?

Should I ask Alberto Martilli? Would that help?

**Comment by dargueso** [comment]
Hi @matthiasdemuzere,

As a general rule, I agree that directly modifying the geo_em files must be done carefully, because there are interdependencies in the variables. In our case, I modify GREENFRAC the same way I do with LU_INDEX. The urban footprint is very clear in GREENFRAC. 

I don't modify LAI12M and ALBEDO12M. There is an urban signature in LAI12M, and ideally this should be modified too, but the original information is very coarse. There is no urban signature for ALBEDO12M, and as far as I understand, ALBEDO12M correspond to the vegetation at each grid cells, including urban grid-cells. The albedo of urban surfaces is prescribed in the URBPARAM.TBL. The final albedo (and I am speculating here) should be a combination of both, taking into consideration the urban fraction and the vegetation at each grid cell. The precise calculation from those two variables to the final albedo, that I don't know.

The other variable I modify is the LANDUSEF, which is different to 0 and 1 if the grid is coarse. In our case, they are all 0 and 1, except for the interpolation thing I showed you this morning. All these variables are in the default geo_em.d01.nc and you can ncview it to get an idea of what each variable looks like.

So, I'd say, we need to change LU_INDEX, LANDUSEF and GREENFRAC in the geo_em.d01.nc, especially to keep consistency with experiment A already running.

[Again, I got your messages while I was writing, it looks like you've been looking at it already]

**Comment by matthiasdemuzere** [comment]
Damn, this is getting much more complicated then I thought. And now I am not sure how we should best proceed with this. Either:
1. use the default way with support from @andreazonato, and focus on our research question only
2. spend time to develop a correctly working new module, and both address our research question AND deliver an easy-to-use tool that might have a large impact on the community.

I am still in favour of 2. but feel insecure because I do not know enough about WRF's pre-processing.

Taking a step back, I have two questions for you @andreazonato:
1. If one uses the default procedure, is the urban extent adjusted to the LCZ urban extent, or is the default urban extent kept so that only LCZ classes are provided to that urban extent?
2. with the default procedure, are all the issues we talk about above addressed? In other words, are all these other fields also adjusted? And if so, that is then done by the WPS system?

If that is the case, and we still want to follow step 2., we probably have to reverse engineer what happens into WPS? That is exactly what we had to do for COSMO as well, something we refered to as the double-counting problem. As also in COSMO many parameters depend on the land-use definition; so by changing this, also these parameters need to be changed. See my code [here](https://github.com/matthiasdemuzere/WUDAPT-to-COSMO/blob/48273c5de2d7fe0da4cce16f4ee9c8efbf779da4/utils.py#L307).

Yet with COSMO I know where to look to do this reverse engineering. For WRF I do not ....
Is this something we can figure out with the 3 of us, or do we need support?

I was example thinking of Ashley Broadbent. He is a WRF expert, a nice guy to work with, and he also implemented the LCZs for our WRF-LCZ tutorials at the BUCSS summer school. 

Happy to hear your thoughts ...

**Comment by matthiasdemuzere** [comment]
And one thing we should keep in mind: we are now focussing on these features for this specific domain. I can imagine that other urban areas show a much more pronounced difference for these GREENFRAC, ALBEDO, ... fields?

**Comment by dargueso** [comment]
@matthiasdemuzere, I would be inclined to option 2 as well, the impact on the community would be much bigger. I'm happy to contribute to this as much as I can. I have used WRF for many years, but I have always dealt with geogrid in a pretty standard way, so not sure how complicated it can get. Also, I have limited time to devote to this, just like you guys, I imagine.

Anyway, on the technical side, as far as I know, Albedo and LAI are coming from MODIS and interpolated into the model grid. That means there isn't a one-to-one relationship with the landuse, although they must be correlated somehow. I ignore the characteristics of the input data from MODIS, but it is relatively coarse. I also asume they're time-variant, since these variables are monthly in the geo_em. In summary, working through standard way and create a flexible routine implies overriding this information somehow. I have no clear way of doing this if we don't have another source of data (maybe you guys do). GREENFRAC is easier I would say, because we can get it from the surrounding areas and the link with land-use is clearer, at least for KL. 

I'm totally open to bringing other people on board, especially if this project is growing from the original idea.

**Comment by matthiasdemuzere** [comment]
Ok. Let's take the plunge, and go for 2. Spend time to develop a correctly working new module, and both address our research question AND deliver an easy-to-use tool that might have a large impact on the community.

Let me try to break this down a bit, for clarity:
**GOAL**: Develop a python package that ingests LCZ information directly into the default geom_em*.nc file, skipping all intermediate steps.

**GENERAL PROCEDURE**: 
- STEP1: Replace the default urban extent with the surrounding naturel land cover
- STEP2: Put back LCZ extent, accompanied by the required additional urban parameters

**QUESTIONs**:
All seem to relate to the following parameters: GREENFRAC, ALBEDO12M, LAI12M, LANDUSEF
- How should they be co-adjusted in both steps?
- How are they depending on the original urban fraction?
- How are they produced in the default WRF pre-processor
- How are they used in the model

@dargueso: perhaps a more specific question for you on LANDUSEF:
To me it is not so clear what the difference is between LANDUSEF and LU_INDEX. As far as I can see, LANDUSEF values are always 1, so a pixel can not have eg. 50% urban (0.5) and 50% Deciduous Broadleaf Forest (0.5)? Does that mean that WRF does not use a mosaic approach? 

If so: 
- that would simplify things already for LU_INDEX and LANDUSEF: when adding the LCZ urban extent, I have to make sure these are consistent.
- But then I am not sure I understand the role of GREENFRAC. This value is > 0 also for urban pixels, but if the LANDUSEF = 0 for all naturel classes, how is GREENFRAC used then?

**Comment by andreazonato** [comment]
Sorry for the late reply guys, I've been busy with external issues.
Replying to your questions:
 _"If one uses the default procedure, is the urban extent adjusted to the LCZ urban extent, or is the default urban extent kept so that only LCZ classes are provided to that urban extent?"_
That's up to the user. I did not modify any of the preprocessing methods. Personally, I remove the default urban extent, and I
put the LCZ extent over the rural landuse.  Others instead put the LCZ over the pre-existing default urban, in order to keep areas that are not described by LCZ with standard urban extent.

_"with the default procedure, are all the issues we talk about above addressed? In other words, are all these other fields also adjusted? And if so, that is then done by the WPS system?"_
In this case, other variables are not modified by LCZ, apart for LU_INDEX and IVGTYP. But in my opinion, for URBAN simulations (i.e. with urban schemes turned on) that's not an issue. In fact, urban schemes act differently with respect to rural landuse schemes, and they do not need any of LAI12M ALBEDO12M and so on. For urban areas, the vegetation fraction is treated separately and the values taken from the tables. Then the surface schemes make a weighted average of the fluxes coming from the urban and the vegetated parts inside the urban grid cells, depending on the urban fraction. So that's not an issue.

The issue, in my opinion, is simply for (in our case) the experiment A). Here LU_INDEX and (as @dargueso  suggests) GREENFRAC should be modified. GREENFRAC should be set to a value similar to the surrounding rural areas, in order to make the replaced urban areas equal to what vegetation would be in those cells.


Hope I've been clear.

**Comment by dargueso** [comment]
I agree with @andreazonato in all his comments.
I would add, answering to  @matthiasdemuzere's question:

> @dargueso: perhaps a more specific question for you on LANDUSEF:
> To me it is not so clear what the difference is between LANDUSEF and LU_INDEX. As far as I can see, LANDUSEF values are always 1, so a pixel can not have eg. 50% urban (0.5) and 50% Deciduous Broadleaf Forest (0.5)? Does that mean that WRF does not use a mosaic approach?

As far as I can tell, at high-res (1km) they take 0 or 1 only. At coarser resolutions (e.g. 10km) they take intermediate values as well. Given that most of the original information is at 30sec resolution (~1km), I would assume that most of each WRF grid cell are covered by a single category and thus the mosaic approach is not visible. On the other hand, it is visible at 10km, where each grid cell likely includes multiple categories. You can check the 10-km geo_em to see that.

Cheers,
Dani

**Comment by andreazonato** [comment]
> I agree with @andreazonato in all his comments.
> I would add, answering to @matthiasdemuzere's question:
> 
> > @dargueso: perhaps a more specific question for you on LANDUSEF:
> > To me it is not so clear what the difference is between LANDUSEF and LU_INDEX. As far as I can see, LANDUSEF values are always 1, so a pixel can not have eg. 50% urban (0.5) and 50% Deciduous Broadleaf Forest (0.5)? Does that mean that WRF does not use a mosaic approach?
> 
> As far as I can tell, at high-res (1km) they take 0 or 1 only. At coarser resolutions (e.g. 10km) they take intermediate values as well. Given that most of the original information is at 30sec resolution (~1km), I would assume that most of each WRF grid cell are covered by a single category and thus the mosaic approach is not visible. On the other hand, it is visible at 10km, where each grid cell likely includes multiple categories. You can check the 10-km geo_em to see that.
> 
> Cheers,
> Dani

Standardly, WRF does not use the Mosaic approach, so the variable LANDUSEF is not used during the simulations. The MOSAIC approach would be really useful instead of doing the aggregation process. But to do it we need to do a bit more of hard-coding in WRF to standardize the process. I'm discussing this with Alberto Martilli, to make the modifications in the 4.4 version

**Comment by matthiasdemuzere** [comment]
That's a lot of useful info @dargueso and @andreazonato.

To summarize and comment:
> That's up to the user. I did not modify any of the preprocessing methods. Personally, I remove the default urban extent, and I
put the LCZ extent over the rural landuse. Others instead put the LCZ over the pre-existing default urban, in order to keep areas that are not described by LCZ with standard urban extent.

Ok. Yet a bit strange in case you have MODIS urban pixels where you do not have LCZ pixels? As you can then not assign an LCZ class to this MODIS urban pixel?

> In this case, other variables are not modified by LCZ, apart for LU_INDEX and IVGTYP. But in my opinion, for URBAN simulations (i.e. with urban schemes turned on) that's not an issue. In fact, urban schemes act differently with respect to rural landuse schemes, and they do not need any of LAI12M ALBEDO12M and so on. For urban areas, the vegetation fraction is treated separately and the values taken from the tables. Then the surface schemes make a weighted average of the fluxes coming from the urban and the vegetated parts inside the urban grid cells, depending on the urban fraction. So that's not an issue.

So no need whatsoever to deal with LAI12M and ALBEDO12M? That is nice. 

> The issue, in my opinion, is simply for (in our case) the experiment A). Here LU_INDEX and (as @dargueso suggests) GREENFRAC should be modified

@dargueso did adjust GREENFRAC based in surrounding pixels, so also this seems fine for Experiment A.


> the Mosaic approach, so the variable LANDUSEF is not used during the simulations. 

If the Mosaic approach is not the default approach, then I suggest to leave LANDUSEF untouched for now. To keep things simple. 


We'll just have to make sure to keep track of all these decisions, that we can clearly describe the do's and don'ts with this tool...

**Comment by matthiasdemuzere** [comment]
So from all of the above, I see the following to do's, per experiment:

**Experiment B:**
- Replace MODIS LCZ with URBAN LCZ
- Make sure to also replace GREENFRAC
- LANDUSEF could be changed for consistency, but will not be used

**Experiment C:**
- Also add LCZ class numbers to LU_INDEX
- Add URB_PARAM matrix to netcdf file.

Is the above correct @dargueso and @andreazonato?

So I could aim straight for the development of Experiment C, as the resulting geo_em*.nc file can also be used for Experiment B, because the use of LCZ info is decided by a namelist flag. Correct @andreazonato ?

**Comment by dargueso** [comment]
Sorry @andreazonato, I mixed up the mosaic approach and the aggregation approach, I think. The resolution rationale still applies to the aggregation, and it results in LANDUSEF values different from 0 and 1 in the coarse resolution domain, as I understand it.

Anyway, I think @matthiasdemuzere plan sound good. The Exp B is definitely a simplification of C, so I don't see a problem on your idea. I would look at LANDUSEF, just in case, to make sure there isn't anything weird, but in the 1k domain it should take 1 and 0 depending on the category, so quite easy to create.

Thanks!

**Comment by matthiasdemuzere** [comment]
For GREENFRAC.

So @dargueso routine replaces MODIS urban with surrounding natural, not only for LU_INDEX but also for GREENFRAC.
Then I put in the LCZ urban extent, and I thus need to adjust GREENFRAC again.

My idea: Tke the GREENFRAC mean / month over all MODIS urban pixels from the default file, and assign that to LCZ urban pixels. Does that make sense?

If not, any other suggestions? @dargueso @andreazonato?

FYI: here are the mean and std (= spatial variability) of GREENFRAC, per month, and over all original MODIS Urban pixels:
- mean: [0.52524465, 0.5226503, 0.5259224, 0.5437781, 0.55109507, 0.55629987, 0.5701588, 0.5710257, 0.55504775, 0.5377472, 0.5298171, 0.5249923]
- std: [0.14393537, 0.14324147, 0.14215636, 0.14097373, 0.14101075, 0.14269516, 0.14676239, 0.1480383, 0.14668752, 0.14597473, 0.14583844, 0.14585523]

As you can see, there is quite a bit of variability over the urban extent. Which I completely remove using my procedure ....

But, then again. Since there is no mosaic approach, and urban parameters are taken from look-up tables, perhaps GREENFRAC is also not used for urban pixels?

**Comment by matthiasdemuzere** [comment]
> So I could aim straight for the development of Experiment C, as the resulting geo_em*.nc file can also be used for Experiment B, because the use of LCZ info is decided by a namelist flag. Correct @andreazonato ?

I believe this will not work.

LU_INDEX is a 2d field, so either:
- has value 13 for urban pixels
- or has values 30, ..., 41 denoting the LCZ class. 

So probably best to make 2 separate outputs?
- One just having the urban extent
- A second one with all LCZ information 

Having both, a user can still decide what experiment to run. Even though I assume most users using this tool will be interested in the settings we ourselves use for Experiment C?

**Comment by andreazonato** [comment]
> So from all of the above, I see the following to do's, per experiment:
> 
> **Experiment B:**
> 
> * Replace MODIS LCZ with URBAN LCZ
> * Make sure to also replace GREENFRAC
> * LANDUSEF could be changed for consistency, but will not be used
> 
> **Experiment C:**
> 
> * Also add LCZ class numbers to LU_INDEX
> * Add URB_PARAM matrix to netcdf file.
> 
> Is the above correct @dargueso and @andreazonato?
> 
> So I could aim straight for the development of Experiment C, as the resulting geo_em*.nc file can also be used for Experiment B, because the use of LCZ info is decided by a namelist flag. Correct @andreazonato ?

Yes, I agree with you with this methodology. That should work properly and should give increasing better results in terms of comparison with observational data.

**Comment by andreazonato** [comment]
> > So I could aim straight for the development of Experiment C, as the resulting geo_em*.nc file can also be used for Experiment B, because the use of LCZ info is decided by a namelist flag. Correct @andreazonato ?
> 
> I believe this will not work.
> 
> LU_INDEX is a 2d field, so either:
> 
> * has value 13 for urban pixels
> * or has values 30, ..., 41 denoting the LCZ class.
> 
> So probably best to make 2 separate outputs?
> 
> * One just having the urban extent
> * A second one with all LCZ information
> 
> Having both, a user can still decide what experiment to run. Even though I assume most users using this tool will be interested in the settings we ourselves use for Experiment C?

The flag is used in order to extend the urban landuse from 3 (or simply one) to the 11 LCZs. The difference between experiment B) and C) is the presence (or not) of the URB_PARAM matrix, in order to evaluate the difference between the majority value (B) and  the aggregation process (C). Regarding the GREENFRAC replacement, I agree with your methodology @matthiasdemuzere. Anyway, values of GREENFRAC for rural areas are pretty similar, and (in my opinion) sensitivity to this variable is really low.

**Comment by matthiasdemuzere** [comment]
> The flag is used in order to extend the urban landuse from 3 (or simply one) to the 11 LCZs. The difference between experiment B) and C) is the presence (or not) of the URB_PARAM matrix, in order to evaluate the difference between the majority value (B) and the aggregation process (C).

Ok. But, my idea was to run our experiment B. (=only one urban class) with the LU_INDEX values that contain LCZ numbers (30-41). That can't work no? If one sets the flag to use to one urban class, I expect this class is defined by LU_INDEX == 13?

**Comment by andreazonato** [comment]
> > The flag is used in order to extend the urban landuse from 3 (or simply one) to the 11 LCZs. The difference between experiment B) and C) is the presence (or not) of the URB_PARAM matrix, in order to evaluate the difference between the majority value (B) and the aggregation process (C).
> 
> Ok. But, my idea was to run our experiment B. (=only one urban class) with the LU_INDEX values that contain LCZ numbers (30-41). That can't work no? If one sets the flag to use to one urban class, I expect this class is defined by LU_INDEX == 13?

No, it won't work. To run the experiment B) with a single urban class, every 30-41 class should be replaced with the 13th class (and even IVGTYP). Another possibility is to run it with the 30-41 urban classes, but with geometric and thermal urban parameteres taken from the lookup table (instead of reading them from the URB_PARAM matrix and then avoiding the aggregation process).

**Comment by matthiasdemuzere** [comment]
> Regarding the GREENFRAC replacement, I agree with your methodology @matthiasdemuzere. Anyway, values of GREENFRAC for rural areas are pretty similar, and (in my opinion) sensitivity to this variable is really low.

Ok. But it seems this Kuala Lumpur area is rather specific. Have a closer look at GREENFRAC from the original geo_em.nc file (same one as posted before).

![image](https://user-images.githubusercontent.com/6926916/115454385-049dc100-a221-11eb-85ef-adc50a12b121.png)

You can see lower than average values for the urban area, **AND** the areas around that (all LU_INDEX == 12). So our GREENFRAC values we sample from surrounding pixels is generally lower than the rural average (if you look at the whole domain). I guess this is just an assumption we have to be clear about?

And the figure also shows the variability of GREENFRAC within the urban extent (see left panel). Something I also remove.

But if you say the sensitivity to this parameter is low, it seems ok to make these assumptions ....

**Comment by matthiasdemuzere** [comment]
> (and even IVGTYP)

A last one for today. @andreazonato you refer to IVGTYP once and a while, but it is not clear to me what variable this is? I do not see it anywhere in the geom file?

**Comment by andreazonato** [comment]
My mistake. IVGTYP appears only in the wrfinput, not in the geoem files. No need to modify it. Sorry

**Comment by dargueso** [comment]
Hey, I have gone through your comments. You're really active! Great to see :)
Going from LCZ LU_INDEX (31-40) to standard urban LU_INDEX (13) can be added as an option in the tool, it is just a matter of replacing all values in the range 31-40 to 13 in the geo_em file. So I'd say, as I think you both suggest, let's prepare everything for exp C, and then we can modify the geo_em to make it suitable for exp B.

I would not worry too much about GREENFRAC either. I don't think using the mean over all urban grid cells will have a significant impact, plus I can't think of a quick alternative.

**Comment by dargueso** [comment]
Ok, so this goes back to something we've discussed already. Sorry about all the confusion.
In my old notes I wrote that because large areas around the city have been deforested for agriculture, replacing GREENFRAC with the dominant surrounding value does not help much, so I decided to keep the original one, which were approximately 0.55. If we want to effectively replace GREENFRAC with the "natural" one, that would involve some assumptions that are not easy to generalise for other cities (e.g. how far out do we look for the natural environment).

In summary, I think keeping GREENFRAC is not ideal, but it is the best we can do in order to generalise the approach (and the tool).  I have updated the code to change GREENFRAC, but we may need to comment this section out and say in the README that GREENFRAC is left unmodified. Happy to discuss.
